### PR TITLE
ESP32 Devcontainer fixes

### PR DESCRIPTION
**Pull request by networkfusion**
## Description
* Adds the GCC path to the ESP32 container
* Fixes the patch version needed
* Improves ability to read the environment variable

## Motivation and Context
 * ESP32 dev container failed to build

## Types of changes

- [ ] Improvement (non-breaking change that improves a feature, code or algorithm)
- [ ] Bug fix (non-breaking change which fixes an issue with code or algorithm)
- [ ] New feature (non-breaking change which adds functionality to code)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Config and build (change in the configuration and build system, has no impact on code or features)
- [x] Dev Containers (changes related with Dev Containers, has no impact on code or features)
- [ ] Dependencies (update dependencies and changes associated, has no impact on code or features)
- [ ] Documentation (changes or updates in the documentation, has no impact on code or features)

## Checklist

- [ ] My code follows the code style of this project (only if there are changes in source code).
- [ ] My changes require an update to the documentation (there are changes that require the docs website to be updated).
- [ ] I have updated the documentation accordingly (the changes require an update on the docs in this repo).
- [ ] I have read the [CONTRIBUTING](https://github.com/nanoframework/.github/blob/main/CONTRIBUTING.md) document.
- [ ] I have tested everything locally and all new and existing tests passed (only if there are changes in source code).

### CODE REVIEW

**Comment by alberk8** [comment]
@networkfusion The line break on the PATH does not work as intended.  I have tested it and it failed to build the container.  Still need to remove all **spaces** from the script.  It still looks ugly and I am not an expert on Linux script.  Is there any more elegant way?    The script below works all spaces are removed.

```
ENV PATH=$PATH:\
$IDF_PATH/components/esptool_py/esptool:\
$IDF_PATH/components/espcoredump:\
$IDF_PATH/components/partition_table/:\
$IDF_PATH/tools/:\
$IDF_PATH/components/app_update:\
/root/.espressif/tools/xtensa-esp32-elf/esp-2021r2-patch5-8.4.0/xtensa-esp32-elf/bin:\
/root/.espressif/tools/xtensa-esp32s2-elf/esp-2021r2-patch5-8.4.0/xtensa-esp32s2-elf/bin:\
/root/.espressif/tools/riscv32-esp-elf/esp-2021r2-patch5-8.4.0/riscv32-esp-elf/bin
```

**Comment by networkfusion** [comment]
> 

Did it work when you changed it to use that?

Docker files are a different beast altogether and only presumed it was the same as what was used with YAML. I have also just checked this reference... https://docs.docker.com/develop/develop-images/dockerfile_best-practices/#sort-multi-line-arguments .

However, if what you have done works, it is still a MAJOR improvement on the previous implementation, so would gladly endorse it!

**Comment by alberk8** [comment]
> > 
> 
> Did it work when you changed it to use that?
> 
> Docker files are a different beast altogether and only presumed it was the same as what was used with YAML. I have also just checked this reference... https://docs.docker.com/develop/develop-images/dockerfile_best-practices/#sort-multi-line-arguments .
> 
> However, if what you have done works, it is still a MAJOR improvement on the previous implementation, so would gladly endorse it!

Yes, it does work when all the spaces are removed.

**Comment by networkfusion** [comment]
> 

Cool... We have a solution then. But lets wait for @Ellerbach to provide input (just incase there is a better solution).

**Comment by networkfusion** [comment]
@alberk8 , just as a test, can you move the colons `:` to new lines and keep the spaces to `\` as that might fix it...

**Comment by networkfusion** [comment]
i.e.
```
ENV PATH=$PATH \
:$IDF_PATH/components/esptool_py/esptool \
:$IDF_PATH/components/espcoredump \
:$IDF_PATH/components/partition_table/ \
:$IDF_PATH/tools/ \
:$IDF_PATH/components/app_update \
:/root/.espressif/tools/xtensa-esp32-elf/esp-2021r2-patch5-8.4.0/xtensa-esp32-elf/bin \
:/root/.espressif/tools/xtensa-esp32s2-elf/esp-2021r2-patch5-8.4.0/xtensa-esp32s2-elf/bin \
:/root/.espressif/tools/riscv32-esp-elf/esp-2021r2-patch5-8.4.0/riscv32-esp-elf/bin
```

**Comment by alberk8** [comment]
> i.e.
> 
> ```
> ENV PATH=$PATH \
> :$IDF_PATH/components/esptool_py/esptool \
> :$IDF_PATH/components/espcoredump \
> :$IDF_PATH/components/partition_table/ \
> :$IDF_PATH/tools/ \
> :$IDF_PATH/components/app_update \
> :/root/.espressif/tools/xtensa-esp32-elf/esp-2021r2-patch5-8.4.0/xtensa-esp32-elf/bin \
> :/root/.espressif/tools/xtensa-esp32s2-elf/esp-2021r2-patch5-8.4.0/xtensa-esp32s2-elf/bin \
> :/root/.espressif/tools/riscv32-esp-elf/esp-2021r2-patch5-8.4.0/riscv32-esp-elf/bin
> ```

No go.  any space anywhere would cause the build to fail.

**Comment by networkfusion** [comment]
> > i.e.
> > ```
> > ENV PATH=$PATH \
> > :$IDF_PATH/components/esptool_py/esptool \
> > :$IDF_PATH/components/espcoredump \
> > :$IDF_PATH/components/partition_table/ \
> > :$IDF_PATH/tools/ \
> > :$IDF_PATH/components/app_update \
> > :/root/.espressif/tools/xtensa-esp32-elf/esp-2021r2-patch5-8.4.0/xtensa-esp32-elf/bin \
> > :/root/.espressif/tools/xtensa-esp32s2-elf/esp-2021r2-patch5-8.4.0/xtensa-esp32s2-elf/bin \
> > :/root/.espressif/tools/riscv32-esp-elf/esp-2021r2-patch5-8.4.0/riscv32-esp-elf/bin
> > ```
> 
> No go. any space anywhere would cause the build to fail.

:-( worth a check... but thanks for trying.

**Comment by networkfusion** [comment]
I have just updated the files with your suggested fixes, and also made it easier to manage future changes (I hope I have not broke anything).

Hopefully more input will be provided in the morrow... but time for bed (for me)!

**Review by alberk8 — COMMENTED** [review]
There should be any spaces in the next line (No space before, after and in between)  . for example below, will not work.

` ^^^$IDF_PATH/components/esptool_py/esptool:\`

File: .devcontainer/sources/Dockerfile.All [hunk 1]
```diff
@@ -112,9 +112,19 @@ RUN ln -fs /usr/bin/python3 /usr/bin/python \
 
 # Install ESP-IDF
 ENV IDF_PATH=/sources/esp-idf
+ENV ESP_PATCH_VER=esp-2021r2-patch5-8.4.0
```
Comment (josesimoes) [inline .devcontainer/sources/Dockerfile.All:115]: Extracting this it's a nice improvement.
And the layout improvements below too, makes it more readable and easy to maintain.

File: .devcontainer/sources/Dockerfile.ESP32 [hunk 2]
```diff
@@ -48,8 +48,12 @@ RUN git clone --branch mbedtls-2.28.1 https://github.com/ARMmbed/mbedtls.git --d
 # Clone ESP-IDF
 RUN git clone --branch v4.4.3 https://github.com/espressif/esp-idf --depth 1 --recursive ./sources/esp-idf
 
+# set gcc location
```
Comment (josesimoes) [inline .devcontainer/sources/Dockerfile.ESP32:51]: These aren't doing anything as we're not downloading nor using GCC official distribution in this container.

File: .devcontainer/sources/Dockerfile.ESP32 [hunk 3]
```diff
@@ -48,8 +48,12 @@ RUN git clone --branch mbedtls-2.28.1 https://github.com/ARMmbed/mbedtls.git --d
...
+ENV ARM_GCC_PATH=/usr/local/bin/gcc
+ENV PATH=$ARM_GCC_PATH/bin:${PATH}
+
 # Copy from our other container
-COPY --from=downloader /tmp/dc-extracted/gcc /usr/local/bin/gcc 
+COPY --from=downloader /tmp/dc-extracted/gcc $ARM_GCC_PATH
```
Comment (josesimoes) [inline .devcontainer/sources/Dockerfile.ESP32:56]:
```suggestion
```

File: .devcontainer/sources/Dockerfile.ESP32 [hunk 4]
```diff
@@ -59,9 +63,19 @@ RUN ln -fs /usr/bin/python3 /usr/bin/python \
 
 # Install ESP-IDF
 ENV IDF_PATH=/sources/esp-idf
+ENV ESP_PATCH_VER=esp-2021r2-patch5-8.4.0
```
Comment (josesimoes) [inline .devcontainer/sources/Dockerfile.ESP32:66]: Same here. This is a nice improvement!

**Review by josesimoes — APPROVED** [review]
LGTM now!

**Review by Ellerbach — APPROVED** [review]
If that's been tested and working, looks good to me :-)

**Pull request merged**